osf/management/commands/handle_duplicate_files.py

```diff
@@ -0,0 +1,262 @@
+      GROUP BY (target_object_id, name, parent_id, type, _path)
+    ) as foo
+    WHERE ct > 1;
+    """
+
+# googledrive and trashedfile/folder/folders excluded.
+valid_file_types = [
+    'osf.onedrivefile',
+    'osf.gitlabfile',
+    'osf.dropboxfile',
+    'osf.githubfile',
+    'osf.s3file',
+    'osf.boxfile',
+    'osf.figsharefile',
+    'osf.osfstoragefile',
+    'osf.bitbucketfile',
+    'osf.owncloudfile',
+    FOLDER
+]
+
+def compare_location_and_region(first_file, second_file):
+    """
+    Compares versions for osfstoragefiles -
+    Compares location['bucket'] and location['object'].
+    Also compares version region_ids.
+    """
+    first_loc = first_file['versions__location']
+    second_loc = second_file['versions__location']
+
+    if first_loc and second_loc:
+        return first_loc['bucket'] == second_loc['bucket'] and \
+            first_loc['object'] == second_loc['object'] and \
+            first_file['versions__region_id'] == first_file['versions__region_id']
+    return False
+
+def build_filename_set(files):
+    """
+    Convenience method for building a set of filenames
+    """
+    return set(files.values_list('name', flat=True).order_by('name'))
+
+def compare_basefilenodes(resource_one, resource_two):
+    """
+    Recursively compares the contents of two folders - assumes initial resource_one
+    and resource_two have the same name
+    """
+    match = True
+    item_count = 0
+    files_list_one = BaseFileNode.objects.filter(parent_id=resource_one['id'])
+    files_list_two = BaseFileNode.objects.filter(parent_id=resource_two['id'])
+    if len(files_list_one) != len(files_list_two) or \
+            build_filename_set(files_list_one) != build_filename_set(files_list_two):
+        match = False
+
+    files_list_one = return_basefilenode_values(files_list_one)
+    while match and (item_count < len(files_list_one)):
+        resource_one = files_list_one[item_count]
+        resource_two = return_basefilenode_values(files_list_two.filter(name=resource_one['name'])).first()
+        match = resources_match(resource_one, resource_two) if resource_two else False
+        item_count += 1
+    return match
+
+def resources_match(resource_one, resource_two):
+    """
+    Checks if resource_one and resource_two match.  If two folders, recursively compares contents.
+    If two files, compares versions.
+    """
+    if resource_one['type'] == FOLDER:
+        match = compare_basefilenodes(resource_one, resource_two)
+    else:
+        match = compare_versions(resource_one, resource_two)
+    return match
+
+def compare_history(first_file, second_file):
+    """
+    Compares _history field on files.  Addons in these scenarios
+    are very unlikely to have versions, so we must compare their _history.
+    """
+    first_hist = first_file['_history']
+    second_hist = second_file['_history']
+
+    if first_hist and second_hist:
+        return first_hist == second_hist
+    return False
```

Comment:
Nearly all addon duplicates have no versions attached, but their `_history` is the same, so we can use this.  Similarly, almost all addon duplicates have guids that need to be repointed.